### PR TITLE
[REVIEW][FEATURE][GAWE-53] - Set Reset Password

### PR DESCRIPTION
**Pull request by saddam-satria**
#### Description

> After Clicking password reset request link, user should set reset password by token

#### Ticket Reference(s)

[GAWE-53](https://bluehabit.atlassian.net/browse/GAWE-53)

##### UI Changes

> Screenshot or video of UI-related changes

##### Coverage

> Test coverage screenshot.

N/A

#### Note(s)

> Couldnt testing yet, because have different model on user credential, but user can reset their password, I will testing later

[GAWE-53]: https://bluehabit.atlassian.net/browse/GAWE-53?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/bluehabit/eureka/common/AbstractBaseService.java [hunk 1]
```diff
@@ -1,9 +1,3 @@
-/*
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/common/AbstractBaseService.java:1]: Ini knp di remove ya?

File: src/main/java/com/bluehabit/eureka/component/user/UserVerification.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
+import org.springframework.data.annotation.CreatedDate;
+import org.springframework.data.annotation.LastModifiedDate;
+
+import java.time.OffsetDateTime;
+
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/component/user/UserVerification.java:24]: Pakein annotation aja `@Getter` dan `@Setter` jadi ga perlu setter getter..kecuali diperlukan

File: src/main/java/com/bluehabit/eureka/controller/ResetPasswordController.java [hunk 3]
```diff
@@ -27,6 +27,6 @@ public ResponseEntity<BaseResponse<Object>> resetPassword(
         @RequestHeader(value = "4adf-3ed", required = false) String token,
         @RequestBody ResetPasswordRequest request
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/controller/ResetPasswordController.java:28]: Gabungin aj sama AuthenticationController

File: src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                throw new UnAuthorizedException(HttpStatus.NOT_FOUND.value(), translate("auth.token.invalid"));
+            }
+            final UserVerification userVerified = userVerification.get();
+            final UserCredential userCredential = userVerified.getUser();
+            userCredential.setUserPassword(request.newPassword());
+            userCredentialRepository.save(userCredential);
+
+            return BaseResponse.success(translate("auth.success"), new HashMap<>());
+
+        } catch (UnAuthorizedException exception) {
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java:46]: Ga perlu pake try catch ya..karena otomatis throwable masuk ke errorcontroller

File: src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    private UserVerificationRepository userVerificationRepository;
+    private UserCredentialRepository userCredentialRepository;
+
+    public ResponseEntity<BaseResponse<Object>> reset(String token, ResetPasswordRequest request) {
+        final Optional<UserVerification> userVerification = userVerificationRepository.findByToken(token);
+        try {
+
+            if (userVerification.isEmpty()) {
+                throw new UnAuthorizedException(HttpStatus.NOT_FOUND.value(), translate("auth.token.invalid"));
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java:37]: Klo mau buat selain UnAuthorize pake aja `throw new GeneralErrorException`

**Review by triandamai — CHANGES_REQUESTED** [review]
Cek dulu ya

File: src/main/java/com/bluehabit/eureka/common/AbstractBaseService.java (hunk 1 above)
Comment (saddam-satria) [reply src/main/java/com/bluehabit/eureka/common/AbstractBaseService.java:1]: itu tadi copy dari yg terbaru abstractnya, di grub katanya conflct kan

File: src/main/java/com/bluehabit/eureka/component/user/UserVerification.java [hunk 6]
```diff
@@ -28,34 +34,40 @@
...
-    @Column
-    private String type;
 
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.LAZY, cascade = {CascadeType.DETACH, CascadeType.MERGE, CascadeType.REFRESH, CascadeType.PERSIST})
     @JoinColumn(name = "userId", referencedColumnName = "id")
-    private UserCredential userCredential;
+    private UserCredential user;
 
     @Column
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/component/user/UserVerification.java:59]: ganti pake `@Enumerated(EnumType.ORDINAL)`

File: src/main/java/com/bluehabit/eureka/controller/AuthenticationController.java [hunk 7]
```diff
@@ -31,4 +36,14 @@ public class AuthenticationController {
     public ResponseEntity<BaseResponse<Object>> signUpWithEmail(@NonNull @RequestBody SignUpWithEmailRequest request) {
         return userService.signUpWithEmail(request);
     }
+
+    @PostMapping(
+        path = "/api/v1/auth/reset-password"
+    )
+    public ResponseEntity<BaseResponse<Object>> resetPassword(
+        @RequestHeader(value = "4adf-3ed", required = false) String token,
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/controller/AuthenticationController.java:44]: taro di `Constant` aja ya

File: src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import java.util.HashMap;
+import java.util.Optional;
+
+@Service
+@AllArgsConstructor
+public class ResetPasswordService extends AbstractBaseService {
+
+    private UserVerificationRepository userVerificationRepository;
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java:29]: kasih `@Autowired` ya biar ter inject

File: src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    private UserCredentialRepository userCredentialRepository;
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java:30]: kasih `@Autowired` ya biar ter inject

File: src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+import com.bluehabit.eureka.exception.GeneralErrorException;
+import lombok.AllArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import java.util.HashMap;
+import java.util.Optional;
+
+@Service
+@AllArgsConstructor
```
Comment (triandamai) [inline src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java:26]: ga perlu pake ini ya

**Review by triandamai — CHANGES_REQUESTED** [review]
dikit lagi ya

File: src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java (hunk 8 above)
Comment (saddam-satria) [reply src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java:29]: itu udah ke inject kak, pake lombok

Comment (triandamai) [reply src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java:29]: lombok nya kita pake buat generate Data aja....klo buat dependency injection pake bawaan aja Autowired biar konsisten

Comment (saddam-satria) [reply src/main/java/com/bluehabit/eureka/services/user/ResetPasswordService.java:22]: oke dah saya push ya

**Review by triandamai — APPROVED** [review]

**Pull request merged**